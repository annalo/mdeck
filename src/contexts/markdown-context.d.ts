declare namespace MarkdownContext {
  declare interface ProviderProps {
    children: React.ReactNode;
  }

  declare interface State {
    htmlString: string;
    md: string;
    slideshowLineNumber: number;
    textLineNumber: number;
  }

  const enum ReducerActionType {
    SetHtmlString = "SET_HTML_STRING",
    SetMd = "SET_MD",
    SetSlideshowLineNumber = "SET_SLIDESHOW_LINE_NUMBER",
    SetTextLineNumber = "SET_TEXT_LINE_NUMBER",
  }

  declare type ReducerAction =
    | {
        type: ReducerActionType.SetHtmlString;
        htmlString: string;
      }
    | { type: ReducerActionType.SetMd; md: string }
    | {
        type: ReducerActionType.SetSlideshowLineNumber;
        slideshowLineNumber: number;
      }
    | {
        type: ReducerActionType.SetTextLineNumber;
        textLineNumber: number;
      };
}
