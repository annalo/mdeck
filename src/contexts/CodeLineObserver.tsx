import React, {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from "react";
import * as R from "ramda";

const CODE_LINE_OBSERVER_DEFAULT_INITIAL_STATE = {};

const CodeLineEntriesContext = createContext<
  CodeLineObserver.Entries | undefined
>(undefined);
const CodeLineObserverContext = createContext<
  CodeLineObserver.Observer | undefined
>(undefined);

const CodeLineObserverProvider: React.FC<CodeLineObserver.ProviderProps> = ({
  children,
  initialEntries,
}: CodeLineObserver.ProviderProps) => {
  const [entries, setEntries] = useState(
    initialEntries || CODE_LINE_OBSERVER_DEFAULT_INITIAL_STATE
  );
  const observe: CodeLineObserver.Observe = useCallback(
    (targetEntries) =>
      setEntries((currentEntries) =>
        R.mergeRight(currentEntries, targetEntries)
      ),
    []
  );

  const observerProviderValue = useMemo(() => ({ observe }), [observe]);
  return (
    <CodeLineEntriesContext.Provider value={entries}>
      <CodeLineObserverContext.Provider value={observerProviderValue}>
        {children}
      </CodeLineObserverContext.Provider>
    </CodeLineEntriesContext.Provider>
  );
};

function useCodeLineEntries(): CodeLineObserver.Entries {
  const context = useContext(CodeLineEntriesContext);
  if (context === undefined) {
    throw new Error(
      "useCodeLineEntries must be used within CodeLineObserverProvider or SlideshowObservable"
    );
  }
  return context;
}

function useCodeLineObserver(): CodeLineObserver.Observer {
  const context = useContext(CodeLineObserverContext);
  if (context === undefined) {
    throw new Error(
      "useCodeLineObserver must be used within CodeLineObserverProvider or SlideshowObservable"
    );
  }
  return context;
}

export { CodeLineObserverProvider, useCodeLineEntries, useCodeLineObserver };
