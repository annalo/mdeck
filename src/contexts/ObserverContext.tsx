import React, { createContext } from "react";

interface ObserverProviderProps {
  children: React.ReactElement;
  observer: IntersectionObserver | null;
}

export const ObserverContext = createContext<IntersectionObserver | null>(null);

export const ObserverProvider: React.FC<ObserverProviderProps> = ({
  children,
  observer,
}: ObserverProviderProps) => {
  return (
    <ObserverContext.Provider value={observer}>
      {children}
    </ObserverContext.Provider>
  );
};
