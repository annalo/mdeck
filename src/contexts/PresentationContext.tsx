import React, { createContext, useContext, useState } from "react";

const PRESENTATION_DEFAULT_STATE = false;

const PresentationStateContext = createContext<
  PresentationContext.IsPresented | undefined
>(undefined);
const PresentationActionsContext = createContext<
  PresentationContext.PresentationActions | undefined
>(undefined);

const PresentationContextProvider: React.FC<PresentationContext.ProviderProps> = ({
  children,
  initialState,
}: PresentationContext.ProviderProps) => {
  const [isPresented, setIsPresented] = useState<
    PresentationContext.IsPresented
  >(initialState || PRESENTATION_DEFAULT_STATE);

  const dismiss = () => setIsPresented(false);
  const present = () => setIsPresented(true);
  const presentationActions = { dismiss, present };

  return (
    <PresentationStateContext.Provider value={isPresented}>
      <PresentationActionsContext.Provider value={presentationActions}>
        {children}
      </PresentationActionsContext.Provider>
    </PresentationStateContext.Provider>
  );
};

function usePresentationState(): PresentationContext.IsPresented {
  const context = useContext(PresentationStateContext);
  if (context === undefined) {
    throw new Error(
      "usePresentationState must be used within a PresentationContextProvider"
    );
  }
  return context;
}

function usePresentationActions(): PresentationContext.PresentationActions {
  const context = useContext(PresentationActionsContext);
  if (context === undefined) {
    throw new Error(
      "usePresentationActions must be used within a PresentationContextProvider"
    );
  }
  return context;
}

export {
  PresentationContextProvider,
  usePresentationState,
  usePresentationActions,
};
