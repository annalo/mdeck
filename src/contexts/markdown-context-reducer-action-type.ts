export const enum MarkdownContextReducerActionType {
  SetHtmlString = "SET_HTML_STRING",
  SetMd = "SET_MD",
  SetSlideshowLineNumber = "SET_SLIDESHOW_LINE_NUMBER",
  SetTextLineNumber = "SET_TEXT_LINE_NUMBER",
}
