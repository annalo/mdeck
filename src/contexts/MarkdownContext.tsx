import React, { createContext, useReducer } from "react";
import { parse } from "utils/parse";

interface MarkdownProviderProps {
  children: React.ReactElement[];
}

interface MarkdownContextInitialState {
  activePane: string;
  html: string;
  previewLineNumber: number;
  textLineNumber: number;
  lineNumber: number;
  md: string;
}

const MARKDOWN_CONTEXT_INITIAL_STATE = {
  activePane: "",
  html: "",
  previewLineNumber: 0,
  textLineNumber: 0,
  lineNumber: 0,
  md: "",
};

export const MarkdownContext = createContext<{
  state: MarkdownContextInitialState;
  dispatch: React.Dispatch<any>;
}>({
  state: MARKDOWN_CONTEXT_INITIAL_STATE,
  dispatch: () => null,
});

export const MarkdownProvider: React.FC<MarkdownProviderProps> = ({
  children,
}: MarkdownProviderProps) => {
  function reducer(
    state,
    { type, activePane, textLineNumber, previewLineNumber, lineNumber, md }
  ) {
    switch (type) {
      case "setActivePane":
        return { ...state, activePane };
      case "setTextLineNumber":
        return { ...state, textLineNumber };
      case "setPreviewLineNumber":
        return { ...state, previewLineNumber };
      case "setLineNumber":
        return { ...state, lineNumber };
      case "setMd":
        return { ...state, html: parse(md), md };
      default:
        throw new Error();
    }
  }

  const [state, dispatch] = useReducer(reducer, MARKDOWN_CONTEXT_INITIAL_STATE);
  const context = { state, dispatch };

  return (
    <MarkdownContext.Provider value={context}>
      {children}
    </MarkdownContext.Provider>
  );
};
