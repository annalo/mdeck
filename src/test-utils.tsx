/* eslint-disable import/no-extraneous-dependencies */

import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/Home/ThemeProvider";
import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <SlideObserverProvider>
          <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
        </SlideObserverProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
