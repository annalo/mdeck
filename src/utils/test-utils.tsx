/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/App/ThemeProvider";

import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { PresentationContextProvider } from "contexts/PresentationContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <PresentationContextProvider>
          <SlideObserverProvider>
            <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
          </SlideObserverProvider>
        </PresentationContextProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
