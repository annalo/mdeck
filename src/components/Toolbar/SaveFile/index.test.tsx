import React from "react";
import { saveAs } from "file-saver";

import { fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { render } from "utils/test-utils";

import { MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE } from "contexts/MarkdownContext";

import { SaveFileMenuItem } from ".";

jest.mock("file-saver");

afterEach(() => jest.clearAllMocks());

describe("<SaveFileMenuItem />", () => {
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<SaveFileMenuItem />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("loads the save file form when clicked", () => {
    render(<SaveFileMenuItem />);

    fireEvent.click(screen.getByRole("button"));
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  describe("<SaveFileForm />", () => {
    test("should save file with filename", () => {
      const filename = "test-file";
      const blob = new Blob([MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE.md], {
        type: "text/markdown",
      });

      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      userEvent.type(screen.getByRole("textbox"), `${filename}{enter}`);

      expect(saveAs).toHaveBeenNthCalledWith(1, blob, filename);
    });

    test("should persist last saved filename", () => {
      const filename = "test-file";

      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      userEvent.type(screen.getByRole("textbox"), `${filename}{enter}`);
      fireEvent.click(screen.getByRole("button"));

      expect(screen.getByRole("textbox")).toHaveDisplayValue(filename);
    });

    test("should not persist filename if not saved", () => {
      const filename = "test-file";
      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      userEvent.type(screen.getByRole("textbox"), `${filename}{esc}`);
      fireEvent.click(screen.getByRole("button"));

      expect(screen.getByRole("textbox")).not.toHaveValue();
    });

    test("exits form when form is submitted", () => {
      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      userEvent.type(screen.getByRole("textbox"), "{enter}");

      expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
    });
    test("exits form when cancel button is clicked", () => {
      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      fireEvent.click(screen.getByLabelText(/cancel/i));

      expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
    });
    test("exits form when Escape key is pressed", () => {
      render(<SaveFileMenuItem />);

      fireEvent.click(screen.getByRole("button"));
      userEvent.type(screen.getByRole("textbox"), "{esc}");

      expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
    });
  });
});
