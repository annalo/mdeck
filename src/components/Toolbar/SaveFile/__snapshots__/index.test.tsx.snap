// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveFileMenuItem /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0 3px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 svg {
  fill: #d4d4d4;
  -webkit-transition: 0.4s ease-in-out;
  transition: 0.4s ease-in-out;
}

.c1 svg circle,
.c1 svg path {
  stroke: #888888;
}

.c1 svg:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c1 svg:hover {
  fill: #bababa;
}

.c1 svg:hover circle,
.c1 svg:hover path {
  stroke: #2f2f2f;
}

.c1:hover svg {
  -webkit-transition: opacity 0.4s ease-in-out;
  transition: opacity 0.4s ease-in-out;
  opacity: 0.3;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.c2 {
  position: absolute;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  color: #2f2f2f;
  font-weight: bolder;
  font-size: 8px;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c2:hover {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<li
    aria-label="save-file"
    class="c0"
    role="button"
  >
    <div
      class="c1"
    >
      <svgr-mock />
      <div
        class="c2"
      >
        SAVE
      </div>
    </div>
  </li>
</DocumentFragment>
`;
